test/run.py

```diff
@@ -0,0 +1,248 @@
+    CONTROLLER_PORT, CONTROLLER_TESTS_CONTAINER, CONTROLLER_TESTS_IMAGE, \
+    DATA_DIRECTORY, PLZ_ROOT_DIRECTORY, PLZ_USER, TEST_DIRECTORY, \
+    docker_compose, get_network
+
+
+def start_controller():
+    test_utils.print_info('Building the controller...')
+    test_utils.execute_command([
+        'docker',
+        'image',
+        'build',
+        '--quiet',
+        f'--tag={CONTROLLER_IMAGE}',
+        os.path.join(PLZ_ROOT_DIRECTORY, 'services/controller')])
+
+    test_utils.stop_container(CONTROLLER_CONTAINER)
+
+    test_utils.print_info('Starting the controller...')
+    docker_compose('up', '--quiet-pull', '--detach')
+
+
+def build_cli():
+    test_utils.print_info('Building the CLI...')
+    test_utils.execute_command([
+        'docker',
+        'image',
+        'build',
+        '--quiet',
+        '--target=builder',
+        f'--tag={CLI_BUILDER_IMAGE}',
+        f'--file={os.path.join(PLZ_ROOT_DIRECTORY, "cli", "Dockerfile")}',
+        PLZ_ROOT_DIRECTORY
+    ])
+    test_utils.execute_command([
+        'docker',
+        'image',
+        'build',
+        '--quiet',
+        f'--tag={CLI_IMAGE}',
+        f'--file={os.path.join(PLZ_ROOT_DIRECTORY, "cli", "Dockerfile")}',
+        PLZ_ROOT_DIRECTORY
+    ])
+
+
+def run_controller_tests(network: str, plz_host: str, plz_port: int,
+                         controller_tests_parameters: List[str]) -> bool:
+    test_utils.stop_container(CONTROLLER_TESTS_CONTAINER)
+    test_utils.execute_command([
+        'docker',
+        'image',
+        'build',
+        '--quiet',
+        f'--tag={CONTROLLER_TESTS_IMAGE}',
+        f'--file={os.path.join(TEST_DIRECTORY, "controller", "Dockerfile")}',
+        PLZ_ROOT_DIRECTORY])
+
+    test_utils.print_info('Running controller tests')
+
+    subp = test_utils.execute_command(
+        ['docker',
+         'run',
+         '--name',
+         CONTROLLER_TESTS_CONTAINER,
+         f'--network={network}',
+         f'--env=PLZ_HOST={plz_host}',
+         f'--env=PLZ_PORT={plz_port}',
+         f'--env=PLZ_USER={PLZ_USER}',
+         f'--env=PLZ_PROJECT=controller-tests',
+         CONTROLLER_TESTS_IMAGE,
+         *controller_tests_parameters],
+        fail_on_failure=False)
+    return subp.returncode == 0
+
+
+def get_end_to_end_tests(
+        command_line_specified_tests: Optional[List[str]] = None) -> [str]:
+    if command_line_specified_tests is None \
+            or len(command_line_specified_tests) == 0:
+        end_to_end_tests = []
+        # Run all tests, except in directories named "*.ignored"
+        for tdir in glob(os.path.join(TEST_DIRECTORY, 'end-to-end', '*', '*')):
+            if os.path.isfile(os.path.join(tdir, 'plz.config.json')) \
+                    and '.ignored' not in tdir:
+                end_to_end_tests.append(
+                    # Remove the dir path until end-to-end
+                    tdir[len(os.path.join(TEST_DIRECTORY, '')):])
```

Comment:
This could be a list comprehension.

Reply:
Meh, the condition's too long. When you've got those Intellij suggests to turn the comprehension into a loop, and I think it's right, the list comprehension stops being more readable at a certain complexity of its constituents.

Reply:
Then pull out a function.